### PR TITLE
[MM-48375 Gekidou] Threads Sync Fix

### PR DESCRIPTION
**Pull request by anurag6713**
#### Summary
This PR adds a new table `TeamThreadsSync` to keep track of the `latest` and `earliest` loaded threads to sync without creating any gaps.

* Refactored multiple threads fetching functions into two `syncTeamThreads` & `loadOldThreads`
* Previously the threads loaded through websocket were not shown in Global Threads Screen until they are fetched through rest call to avoid gaps but now all threads are shown if their `last_reply_at` is greater than the `earliest`
* Fixes on entry sync for other servers
* Scroll to the top on the tab change
* Fixes bug when old thread gets replied

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48375
https://mattermost.atlassian.net/browse/MM-43318

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: iOS 15, Android 12

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by enahum** [comment]
@anurag6713 where are we with this?

As this introduces database changes I would've thought is needed before the code freeze

cc: @Willyfrog

**Comment by anurag6713** [comment]
I'm almost done with the syncing. I will test it and then write the DB migration.

File: app/database/migration/server/index.ts
Comment (enahum) [inline app/database/migration/server/index.ts:36]: you could probably use `@ts-expect-error

File: app/database/operator/server_data_operator/handlers/team_threads_sync.ts
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/team_threads_sync.ts:24]:
```suggestion
                'An empty or undefined "data" array has been passed to the handleTeamThreadsSync method',
```

File: app/database/operator/server_data_operator/handlers/thread.test.ts
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/thread.test.ts:198]: I would like to see a test where it updates some records

File: app/actions/websocket/threads.ts
Comment (enahum) [inline app/actions/websocket/threads.ts:14]: why is the await needed here? I don't see this causing any effect

File: app/queries/servers/thread.ts
Comment (enahum) [inline app/queries/servers/thread.ts:40]:
```suggestion
    return result?.[0];
```

File: app/actions/remote/thread.ts
Comment (enahum) [inline app/actions/remote/thread.ts:36]: enums are not great, think of using constants instead

File: app/actions/remote/thread.ts
Comment (enahum) [inline app/actions/remote/thread.ts:344]: I do not quite get the logic here, so I'll defer to @koox00

**Review by enahum — APPROVED** [review]
Most of the code looks good, about the CRT logic I'll defer review to @koox00 
before merging we should test the migration with a bunch of data already loaded.

File: app/actions/websocket/threads.ts
Comment (anurag6713) [reply app/actions/websocket/threads.ts:14]: True, it was unexpectedly added while I was going through all the places this function is used. Reverting it.

**Review by larkox — APPROVED** [review]

File: app/actions/remote/thread.ts
Comment (koox00) [reply app/actions/remote/thread.ts:344]: LGTM

**Comment by enahum** [comment]
@anurag6713 reviews done, any reason why this is still a draft?

**Pull request merged**
